auth/authenticators/token-request-based-authenticator.ts

```diff
@@ -84,15 +84,14 @@ export class TokenRequestBasedAuthenticator extends Authenticator implements Aut
     this.tokenManager.setHeaders(this.headers);
   }
 
-  public authenticate(options: AuthenticateOptions, callback: AuthenticateCallback): void {
-    this.tokenManager.getToken((err, token) => {
-      if (err) {
-        callback(err);
-      } else {
+  public authenticate(options: AuthenticateOptions): Promise<void | Error> {
+    return this.tokenManager.getToken()
+      .then(token => {
         const authHeader = { Authorization: `Bearer ${token}` };
         options.headers = extend(true, {}, options.headers, authHeader);
-        callback(null);
-      }
-    });
+        return;
+      }, err => {
+        throw(err);
```

Comment:
why the `throw`? isn't the caller supposed to handle the `catch()`?

Reply:
All Promise handlers (`then`, `catch`, `finally`) return another Promise. If you return a value inside, it is the same as "resolving" the Promise for the next handler - if you throw an error, it is the same as "rejecting". So as long as the caller anticipates a rejection, this should be fine.

Reply:
I was asking why are you actually doing `,err -> { throw(err);}` I think that if you don't add it then it will be the same

Reply:
Oh, I see what you're saying. You are correct about that, I will update the code